src/photon/photon_algorithm.c

```diff
@@ -419,55 +496,82 @@ void handle_worker_available(local_scheduler_state *state,
-  HASH_ADD(handle, algorithm_state->local_objects, object_id, sizeof(object_id),
-           entry);
-
-  /* If we were previously trying to fetch this object, remove the fetch request
-   * from the hash table. */
-  fetch_object_request *fetch_req;
+  /* Get the entry for this object from the active fetch request, or allocate
+   * one if needed. */
+  object_entry *entry;
   HASH_FIND(hh, algorithm_state->fetch_requests, &object_id, sizeof(object_id),
-            fetch_req);
-  if (fetch_req != NULL) {
-    HASH_DELETE(hh, algorithm_state->fetch_requests, fetch_req);
-    CHECK(event_loop_remove_timer(state->loop, fetch_req->timer) == AE_OK);
-    free(fetch_req);
+            entry);
+  if (entry != NULL) {
+    /* Remove the object from the active fetch requests. */
+    HASH_DELETE(hh, algorithm_state->fetch_requests, entry);
+  } else {
+    /* Allocate a new object entry. Object entries will get freed if the object
+     * is removed. */
+    entry = (object_entry *) malloc(sizeof(object_entry));
+    entry->object_id = object_id;
   }
 
+  /* Add the entry to the set of locally available objects. */
+  HASH_ADD(hh, algorithm_state->local_objects, object_id, sizeof(object_id),
+           entry);
+
   /* Move any tasks whose object dependencies are now ready to the dispatch
    * queue. */
-  /* TODO(swang): This can be optimized by keeping a lookup table from object
-   * ID to list of dependent tasks in the waiting queue. */
-  task_queue_entry *elt, *tmp;
-  DL_FOREACH_SAFE(algorithm_state->waiting_task_queue, elt, tmp) {
-    if (can_run(algorithm_state, elt->spec)) {
-      LOG_DEBUG("Moved task to dispatch queue");
-      DL_DELETE(algorithm_state->waiting_task_queue, elt);
-      DL_APPEND(algorithm_state->dispatch_task_queue, elt);
+  object_dependency *elt;
+  HASH_FIND(dependency_handle, algorithm_state->object_dependencies, &object_id,
+            sizeof(object_id), elt);
+  if (elt != NULL) {
+    HASH_DELETE(dependency_handle, algorithm_state->object_dependencies, elt);
+    task_queue_entry *task_entry = NULL;
+    for (task_queue_entry **p =
+             (task_queue_entry **) utarray_front(elt->dependent_tasks);
+         p != NULL;
+         p = (task_queue_entry **) utarray_next(elt->dependent_tasks, p)) {
+      task_queue_entry *task_entry = *p;
+      if (can_run(algorithm_state, task_entry->spec)) {
+        LOG_DEBUG("Moved task to dispatch queue");
+        DL_DELETE(algorithm_state->waiting_task_queue, task_entry);
+        DL_APPEND(algorithm_state->dispatch_task_queue, task_entry);
+      }
     }
-  }
+    utarray_free(elt->dependent_tasks);
+    free(elt);
 
-  /* Try to dispatch tasks, since we may have added some from the waiting
-   * queue. */
-  dispatch_tasks(state, algorithm_state);
+    /* Try to dispatch tasks, since we may have added some from the waiting
+     * queue. */
+    dispatch_tasks(state, algorithm_state);
+  }
 }
 
 void handle_object_removed(local_scheduler_state *state,
```

Comment:
On the other hand, maybe batching would introduce a big pause in the local scheduler..